src/Streamly/Internal/Data/Fold/Types.hs

```diff
@@ -799,32 +812,51 @@ runStep _ _ = undefined
 --
 {-# INLINE many #-}
 many :: Monad m => Fold m b c -> Fold m a b -> Fold m a c
-many (Fold fstep finitial fextract) (Fold step1 initial1 extract1) =
+many (Fold cstep cinitial cextract) (Fold sstep sinitial sextract) =
     Fold step initial extract
 
     where
 
-    initial = Tuple' <$> initial1 <*> finitial
+    -- cs = collect state
+    -- ss = split state
+    -- cres = collect state result
+    -- sres = split state result
+    -- cb = collect done
+    -- sb = split done
+
+    -- Caution! There is mutual recursion here, inlining the right functions is
+    -- important.
+
+    {-# INLINE handleSplitStep #-}
+    handleSplitStep cs sres =
+        case sres of
+            Partial ss1 -> return $ Partial $ Tuple' ss1 cs
+            Done sb -> loopTillPartial cs sb
+
+    {-# INLINE handleCollectStep #-}
+    handleCollectStep cres =
+        case cres of
+            Partial cs -> do
+                sres <- sinitial
+                handleSplitStep cs sres
+            Done cb -> return $ Done cb
+
+    -- Do not inline this
+    loopTillPartial cs sb = cstep cs sb >>= handleCollectStep
```

Comment:
Rename `loopTillPartial` to something like `runCollector` or something?

Reply:
as you wish, I am neutral.